designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+  * Flags would have to be provided by the user as a part of project configuration.
+
+* Providing other options (other than stdin stdout) for inter-process communication between Kubebuilder and the plugins.
+
+### Examples
+
+* `kubebuilder create api --plugins=myexternalplugin/v1`
+  * should scaffold files using the source code of the external plugin as defined in its implementation of the create api method.
+
+* `kubebuilder create api --plugins=myexternalplugin/v1,myexternalplugin/v2`
+  * should scaffold files using the source code of the external plugin as defined in their implementation of the create api method(by respecting the plugin chaining order, i.e. in the order of create api of v1 and then create api of v2 as specified in the layout field in the configuration).
+
+* `kubebuilder create api --plugins=myexternalplugin/v1 --help`
+  * should provide help implementation of the plugin which is not shipped in the binary (source code is present outside of the KB repo).
+
+### How should Kubebuilder discover plugin binaries
+
+Main goal of phase 2 plugins is discovery of out of tree plugins that are external to Kubebuilder and not shipped with the CLI. To be able to run these plugins via the Kubebuilder CLI, it’s necessary for kubebuilder to discover these external plugin binaries and use them as internal ones. I will discuss a few approaches in the order of suitability and preference, with their respective pros and cons.
+
+#### Approach #1
+
+User specified file paths based approach: User will provide a list of file paths for Kubebuilder to discover the plugins in. We will define a variable `KUBEBUILDER_PLUGINS_DIRS` that will take a list of file paths to search for the plugin name. It will also have a default value to search in, in case no file paths are provided. It will search for the plugin name that was provided to the `--plugins` flag in the CLI. Kubebuilder will recursively search for all file paths until the plugin name is found and returns the successful match, and if it doesn’t exist, it returns an error message that the plugin is not found in the provided file paths.
+
+* Pros
+  * This provides flexibility for the user to specify the file paths that the plugin would be placed in and Kubebuilder could discover the binaries in those user specified file paths.
+
+  * No constraints on plugin binary naming or directory placements from the Kubebuilder side.
+
+  * Provides a default value for the plugin directory in case user wants to use that to drop their plugins.
+
+* Cons
+  * Adds some complexity as we need to have an extra flag to have the user provide those file paths and/or make PROJECT file changes to add an extra field to `plugins` field that is a comma separated list of file paths for every plugin that is defined in the `layout` field.
+
+#### Approach #2
+
+This is how [kustomize](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/) discovers plugins.
```

Comment:
I vote for this approach because this is a common way to configure plugins. See [setup-envtest](https://github.com/kubernetes-sigs/controller-runtime/tree/master/tools/setup-envtest) too.

Reply:
That seems the best I think too. 
I'd recommend we keep this one and add the others in an alternative section just to keep the history. (by following the templated suggested above)

Reply:
Sounds good to me as well. I kept this one in the main section and moved other approaches to Alternatives section.